ec/src/hashing/curve_maps/elligator2/mod.rs

```diff
@@ -0,0 +1,268 @@
+    /// are represented as [-(q-1)/2, (q-1)/2]
+    const Z: Self::BaseField;
+}
+
+/// Represents the Elligator2 hash-to-curve map defined by `P`.
+pub struct Elligator2Map<P: TECurveConfig>(PhantomData<fn() -> P>);
+
+impl<P: Elligator2Config> MapToCurve<Projective<P>> for Elligator2Map<P> {
+    /// Constructs a new map if `P` represents a valid map.
+    fn new() -> Result<Self, HashToCurveError> {
+        // Verifying that U is a non-square
+        if P::Z.legendre().is_qr() {
+            return Err(HashToCurveError::MapToCurveError(
+                "Z should be a quadratic non-residue for the Elligator2 map".to_string(),
+            ));
+        }
+
+        // We assume that the Montgomery curve is correct and  as such we do
+        // not verify the prerequisite for applicability of Elligator2 map
+        Ok(Elligator2Map(PhantomData))
+    }
+
+    /// Map an arbitrary base field element `element` to a curve point.
+    /// Based on
+    /// <https://github.com/zcash/pasta_curves/blob/main/src/hashtocurve.rs>.
+    fn map_to_curve(&self, element: P::BaseField) -> Result<Affine<P>, HashToCurveError> {
+        // 1. x1 = -(J / K) * inv0(1 + Z * u^2)
+        // 2. If x1 == 0, set x1 = -(J / K)
+        // 3. gx1 = x1^3 + (J / K) * x1^2 + x1 / K^2
+        // 4. x2 = -x1 - (J / K)
+        // 5. gx2 = x2^3 + (J / K) * x2^2 + x2 / K^2
+        // 6. If is_square(gx1), set x = x1, y = sqrt(gx1) with sgn0(y) == 1.
+        // 7. Else set x = x2, y = sqrt(gx2) with sgn0(y) == 0.
+        // 8. s = x * K
+        // 9. t = y * K
+        // 10. return (s, t)
+
+        // ark a is irtf J
+        // ark b is irtf k
+        let j = <P as MontCurveConfig>::COEFF_A;
+        let k = <P as MontCurveConfig>::COEFF_B;
+        let j_on_k = j / k;
+
+        let ksq_inv = k
+            .square()
+            .inverse()
+            .expect("B coefficient cannot be zero in Montgomery form");
+
+        let den_1 = <P::BaseField as One>::one() + P::Z * element.square();
+
+        let x1 = -j_on_k
+            / (if den_1.is_zero() {
+                <P::BaseField as One>::one()
+            } else {
+                den_1
+            });
+        let x1sq = x1.square();
+        let x1cb = x1sq * x1;
+        let gx1 = x1cb + j_on_k * x1sq + x1 * ksq_inv;
+
+        let x2 = -x1 - j_on_k;
+        let x2sq = x2.square();
+        let x2cb = x2sq * x2;
+        let gx2 = x2cb + j_on_k * x2sq + x2 * ksq_inv;
+
+        let (x, y, _sig0_y): (P::BaseField, P::BaseField, i8) = if gx1.legendre().is_qr() {
+            (
+                x1,
+                gx1.sqrt()
+                    .expect("We have checked that gx1 is a quadratic residue. Q.E.D"),
+                1,
+            )
+        } else {
+            (
+                x2,
+                gx2.sqrt()
+                    .expect("gx2 is a quadratic residue because gx1 is not. Q.E.D"),
+                0,
+            )
+        };
```

Comment:
```suggestion
        let (x, y): (P::BaseField, P::BaseField) = if gx1.legendre().is_qr() {
            (
                x1,
                gx1.sqrt()
                    .expect("We have checked that gx1 is a quadratic residue. Q.E.D"),
            )
        } else {
            (
                x2,
                gx2.sqrt()
                    .expect("gx2 is a quadratic residue because gx1 is not. Q.E.D"),
            )
        };
```

Reply:
Actually that's only half-right. It seems that parts of steps 6,7 are missing:
```
        // 6. If is_square(gx1), set x = x1, y = sqrt(gx1) with sgn0(y) == 1.
        // 7. Else set x = x2, y = sqrt(gx2) with sgn0(y) == 0.
```
In fact, when I place an assert in the code:
`        assert_eq!(parity(&y), sgn0, "parity of y is not equal to sgn0(y)");`
both tests fail :(

I think that what we need is to return `sgn0`, then call `parity(&y)` and if they mismatch, negate `y`.